File: src/sp/it/pl/gui/itemnode/FieldedPredicateChainItemNode.java
package sp.it.pl.gui.itemnode;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;
import java.util.function.Supplier;
import java.util.stream.Stream;
import sp.it.pl.gui.itemnode.FieldedPredicateItemNode.PredicateData;
import sp.it.pl.util.access.fieldvalue.ObjectField;
import sp.it.pl.util.functional.Functors.Ƒ1;
import static sp.it.pl.util.functional.Util.IS;

/**
 * Chained filter node producing {@link sp.it.pl.util.access.fieldvalue.ObjectField} predicate.
 */
public class FieldedPredicateChainItemNode<T, F extends ObjectField<T,Object>> extends ChainValueNode<Predicate<T>,FieldedPredicateItemNode<T,F>> {

	protected Supplier<PredicateData<F>> supplier;
	protected final List<PredicateData<F>> data = new ArrayList<>();

	public FieldedPredicateChainItemNode(Ƒ1<FieldedPredicateChainItemNode<T,F>,FieldedPredicateItemNode<T,F>> chainedFactory) {
		super(0, Integer.MAX_VALUE, null);
		this.chainedFactory = () -> chainedFactory.apply(this);
		inconsistent_state = false;
		generateValue();
	}

	public List<PredicateData<F>> getData() {
		return data;
	}

	public void setData(List<PredicateData<F>> data) {
		inconsistent_state = true;
		this.data.clear(); // causes serious problems, unknown
		this.data.addAll(data);
		chain.forEach(g -> g.chained.setData(data));
		clear(); // bug fix, not sure if it does not cause problems
	}

	public void setPrefTypeSupplier(Supplier<PredicateData<F>> supplier) {
		this.supplier = supplier;
		chain.forEach(g -> g.chained.setPrefTypeSupplier(supplier));
	}

	public Supplier<PredicateData<F>> getPrefTypeSupplier() {
		return supplier;
	}

	public boolean isEmpty() {
		return chain.stream().allMatch(c -> c.chained.isEmpty());
	}

	public void clear() {
		inconsistent_state = true;
		if (!chain.isEmpty()) chain.setAll(chain.get(0));   // TODO: handle properly
		chain.forEach(c -> c.chained.clear());
		inconsistent_state = false;
		generateValue();
	}

	@Override
	protected Predicate<T> reduce(Stream<Predicate<T>> values) {
		return values.reduce(Predicate::and).orElse((Predicate) IS);
	}

}

File: src/player/main/sp/it/pl/ui/itemnode/FieldedPredicateChainItemNode.kt
package sp.it.pl.ui.itemnode

import javafx.beans.value.WritableValue
import javafx.event.EventHandler
import javafx.scene.Node
import javafx.scene.input.KeyCode.ESCAPE
import javafx.scene.input.KeyCode.F
import javafx.scene.input.KeyEvent
import sp.it.pl.ui.itemnode.FieldedPredicateItemNode.PredicateData
import sp.it.util.access.fieldvalue.ObjectField
import sp.it.util.functional.Util.IS
import sp.it.util.functional.asIs
import sp.it.util.ui.isAnyParentOf
import java.lang.Integer.MAX_VALUE
import java.util.ArrayList
import java.util.function.BiPredicate
import java.util.function.Predicate
import java.util.function.Supplier
import java.util.stream.Stream
import kotlin.streams.asSequence

/** [ObjectField] [Predicate] chain. */
open class FieldedPredicateChainItemNode<T, F: ObjectField<T, Any?>>: ChainValueNode<Predicate<T>, FieldedPredicateItemNode<T, F>, Predicate<T>> {

   constructor(chainedFactory: (FieldedPredicateChainItemNode<T, F>) -> FieldedPredicateItemNode<T, F>): super(0, MAX_VALUE, IS.asIs(), null) {
      this.data = ArrayList()
      this.chainedFactory = Supplier { chainedFactory(this) }
      isHomogeneousRem = BiPredicate { i, _ -> i!=0 }
      isHomogeneousAdd = BiPredicate { _, _ -> true }
      isHomogeneousOn = BiPredicate { _, _ -> true }
      isHomogeneousEdit = BiPredicate { _, _ -> true }
      inconsistentState = false
      generateValue()
   }

   constructor(): this({
      FieldedPredicateItemNode<T, F>().apply {
         setPrefTypeSupplier(it.prefTypeSupplier)
         setData(it.data)
      }
   })

   var prefTypeSupplier: Supplier<PredicateData<F>>? = null
      set(supplier) {
         field = supplier
         chain.forEach { it.chained.setPrefTypeSupplier(supplier) }
      }

   var data: List<PredicateData<F>>
      set(data) {
         inconsistentState = true
         field = data
         chain.forEach { it.chained.setData(data) }
         clear()  // bug fix, not sure if it does not cause problems
      }

   fun isEmpty(): Boolean = chain.all { it.chained.isEmpty }

   /**
    * Shrinks the chain to 1 if [length] > 0, or does not change length otherwise.
    * Invokes [FieldedPredicateItemNode.clear] for each remaining link.
    * Invokes [generateValue] exactly once.
    */
   fun clear() {
      inconsistentState = true
      if (length()>1) convergeTo(1);
      chain.forEach { it.chained.clear() }
      inconsistentState = false
      generateValue()
   }

   override fun reduce(values: Stream<Predicate<T>>): Predicate<T> = values.asSequence().fold(IS.asIs(), Predicate<T>::and)

   fun buildToggleOnKeyHandler(filterVisible: WritableValue<Boolean>, owner: Node) = EventHandler<KeyEvent> { e ->
      when (e.code) {
         F -> {
            if (e.isShortcutDown) {
               if (filterVisible.value) {
                  val hasFocus = root.scene?.focusOwner?.let { root.isAnyParentOf(it) }==true
                  if (hasFocus) {
                     filterVisible.value = !filterVisible.value
                     if (!filterVisible.value) owner.requestFocus()
                  } else {
                     focus()
                  }
               } else {
                  filterVisible.value = true
               }
               e.consume()
            }
         }
         ESCAPE -> {
            if (filterVisible.value) {
               if (isEmpty()) {
                  filterVisible.value = false
                  owner.requestFocus()
               } else {
                  clear()
               }
               e.consume()
            }
         }
         else -> {}
      }
   }

}